contracts/access/AccessControl.sol

```diff
@@ -159,23 +159,24 @@ abstract contract AccessControl is Context {
     /**
      * @dev Grants `role` to `account`.
      *
-     * Emits a {RoleGranted} event.
+     * If `account` had not been already granted `role`, emits a {RoleGranted}
+     * event.
      */
     function _grantRole(bytes32 role, address account) internal virtual {
-        _roles[role].members.add(account);
-
-        emit RoleGranted(role, account, msg.sender);
+        if (_roles[role].members.add(account)) {
```

Comment:
I should've noted before, I'm not sure if this one should be conditional. My understanding is that this sort of "ratification" can be useful in some scenarios, and that in the future nodes can begin trimming down their state so that past logs are not necessarily present. This seems irrelevant for auditing purposes, though, it would only be useful for some sort of UI. And even then, getting the contract to re-emit this event would require contacting an admin and having them make a transaction. Not sure how practical any of this is.

The asymmetry with making only one of granting/revoking conditional does make me doubt though. What do you think?

--- 

As a side note, I think the meaning of the boolean return value of `EnumerableSet.add` is not obvious and it's a place where I will always look out for bugs.

Reply:
I dislike having that asymmetry between `grant` and `revoke` very much, I think it'd be very confusing, and we'd be presuming too much about how `AccessControl` will be used.

Given that roles can be enumerated, not emitting events when the action was a no-op is fine by me (though I can see how both approaches, always emitting and emitting conditionally can be confusing in their own way). But I really think we should use the same guideline for both `grant` and `revoke`.

re. the boolean, what do you find confusing about it?

Reply:
Sounds good to keep both conditional.

The boolean could mean both "true if was already present" or "true if was added".

Reply:
FWIW I got the meaning behind the boolean from [rust](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.insert), all similar functions return true when the call actually modified the underlying data (in other words, false is a no-op).